src/folding/protogalaxy/folding.rs

```diff
@@ -0,0 +1,622 @@
+                return Err(Error::NotSameLength(
+                    "wj.w.len()".to_string(),
+                    wj.w.len(),
+                    "n".to_string(),
+                    n,
+                ));
+            }
+            ws.push(wj.w.clone());
+        }
+
+        let k = vec_instances.len();
+        let H =
+            GeneralEvaluationDomain::<C::ScalarField>::new(k + 1).ok_or(Error::NewDomainFail)?;
+        // WIP review t/d
+        let EH = GeneralEvaluationDomain::<C::ScalarField>::new(t * k + 1)
+            .ok_or(Error::NewDomainFail)?;
+        let L_X: Vec<DensePolynomial<C::ScalarField>> = lagrange_polys(H);
+
+        // K(X) computation in a naive way, next iterations will compute K(X) as described in Claim
+        // 4.5 of the paper.
+        let mut G_evals: Vec<C::ScalarField> = vec![C::ScalarField::zero(); EH.size()];
+        for (hi, h) in EH.elements().enumerate() {
+            // each iteration evaluates G(h)
+            // inner = L_0(x) * w + \sum_k L_i(x) * w_j
+            let mut inner: Vec<C::ScalarField> = vec![C::ScalarField::zero(); ws[0].len()];
+            for (i, w) in ws.iter().enumerate() {
+                // Li_w = Li(X) * wj
+                let mut Li_w: Vec<DensePolynomial<C::ScalarField>> =
+                    vec![DensePolynomial::<C::ScalarField>::zero(); w.len()];
+                for (j, wj) in w.iter().enumerate() {
+                    let Li_wj = &L_X[i] * *wj;
+                    Li_w[j] = Li_wj;
+                }
+                // Li_w_h = Li_w(h) = Li(h) * wj
+                let mut Liw_h: Vec<C::ScalarField> = vec![C::ScalarField::zero(); w.len()];
+                for (j, _) in Li_w.iter().enumerate() {
+                    Liw_h[j] = Li_w[j].evaluate(&h);
+                }
+
+                for j in 0..inner.len() {
+                    inner[j] += Liw_h[j];
+                }
+            }
+            let f_ev = eval_f(r1cs, &inner)?;
+
+            let mut Gsum = C::ScalarField::zero();
+            for (i, f_ev_i) in f_ev.iter().enumerate() {
+                let pow_i_betas = pow_i(i, &betas_star);
+                let curr = pow_i_betas * f_ev_i;
+                Gsum += curr;
+            }
+            G_evals[hi] = Gsum;
+        }
+        let G_X: DensePolynomial<C::ScalarField> =
+            Evaluations::<C::ScalarField>::from_vec_and_domain(G_evals.clone(), EH).interpolate();
+        let Z_X: DensePolynomial<C::ScalarField> = H.vanishing_polynomial().into();
+        // K(X) = (G(X)- F(alpha)*L_0(X)) / Z(X)
+        let L0_e = &L_X[0] * F_alpha; // L0(X)*F(a) will be 0 in the native case
+        let G_L0e = &G_X - &L0_e;
+        // Pending optimization: move division by Z_X to the prev loop
+        let (K_X, remainder) = G_L0e.divide_by_vanishing_poly(H).ok_or(
+            // using Error::ShouldNotReach because the method divide_by_vanishing_poly should never
+            // return 'None' despite returning an 'Option'.
```

Comment:
Similarly to https://github.com/privacy-scaling-explorations/folding-schemes/pull/37#discussion_r1427278046 , the idea for returning this error was to avoid using a panic and then having services crashing in case that this condition is reached, because despite the method `divide_by_vanishing_poly` from arkworks (https://github.com/arkworks-rs/algebra/tree/66ca8ed0f88955c573c077da9134f2db29e18c19/poly/src/polynomial/univariate/dense.rs#L170) should never return `None`, since it has the `Option` it could happen, so I thought about returning that 'warning' error so we can track it.

But now I had a second thought on it, and better to return a proper error so we don't depend on a concrete behaviour from the arkworks method (eg. in case that the underlying arkworks lib changes the behaviour and leads to a case where this error would be triggered).
Updated in https://github.com/privacy-scaling-explorations/folding-schemes/pull/37/commits/0e892f31d65bd4980bdb86d99ed57072f354cfc5 .